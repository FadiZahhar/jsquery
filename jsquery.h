/*-------------------------------------------------------------------------
 *
 * jsquery.h
 *     Definitions of jsquery datatype
 *
 * Copyright (c) 2014, PostgreSQL Global Development Group
 * Author: Teodor Sigaev <teodor@sigaev.ru>
 *
 * IDENTIFICATION
 *    contrib/jsquery/jsquery.h
 *
 *-------------------------------------------------------------------------
 */

#ifndef __JSQUERY_H__
#define __JSQUERY_H__

#include "access/gin.h"
#include "fmgr.h"
#include "utils/numeric.h"
#include "utils/jsonb.h"

typedef struct
{
	int32	vl_len_;	/* varlena header (do not touch directly!) */
} JsQuery;

#define DatumGetJsQueryP(d)	((JsQuery*)DatumGetPointer(PG_DETOAST_DATUM(d)))
#define PG_GETARG_JSQUERY(x)	DatumGetJsQueryP(PG_GETARG_DATUM(x))
#define PG_RETURN_JSQUERY(p)	PG_RETURN_POINTER(p)

typedef enum JsQueryItemType {
		jqiNull = jbvNull, 
		jqiString = jbvString, 
		jqiNumeric = jbvNumeric, 
		jqiBool = jbvBool, 
		jqiArray = jbvArray, 
		jqiAnd = '&', 
		jqiOr = '|', 
		jqiNot = '!',
		jqiEqual = '=',
		jqiLess  = '<',
		jqiGreater  = '>',
		jqiLessOrEqual = '{',
		jqiGreaterOrEqual = '}',
		jqiContains = '@',
		jqiContained = '^',
		jqiOverlap = 'O',
		jqiAny = '*',
		jqiAnyArray = '#',
		jqiAnyKey = '%',
		jqiKey = 'K',
		jqiCurrent = '$',
		jqiIn = 'I'
} JsQueryItemType;

/*
 * Support functions to parse/construct binary value.
 * Unlike many other representation of expression the first/main
 * node is not an operation but left operand of expression. That 
 * allows to implement cheep follow-path descending in jsonb
 * structure and then execute operator with right operand which 
 * is always a constant. 
 */

typedef struct JsQueryItem {
	JsQueryItemType	type;
	int32			nextPos;
	char			*base;

	union {
		struct {
			char		*data;  /* for bool, numeric and string/key */
			int			datalen; /* filled only for string/key */
		} value; 

		struct {
			int32	left;
			int32	right;
		} args;

		int32		arg;

		struct {
			int		nelems;
			int		current;
			int32	*arrayPtr;
		} array;
	};


} JsQueryItem;

extern void jsqInit(JsQueryItem *v, JsQuery *js);
extern void jsqInitByBuffer(JsQueryItem *v, char *base, int32 pos);
extern bool jsqGetNext(JsQueryItem *v, JsQueryItem *a);
extern void jsqGetArg(JsQueryItem *v, JsQueryItem *a);
extern void jsqGetLeftArg(JsQueryItem *v, JsQueryItem *a);
extern void jsqGetRightArg(JsQueryItem *v, JsQueryItem *a);
extern Numeric	jsqGetNumeric(JsQueryItem *v);
extern bool		jsqGetBool(JsQueryItem *v);
extern char * jsqGetString(JsQueryItem *v, int32 *len);
extern void jsqIterateInit(JsQueryItem *v);
extern bool jsqIterateArray(JsQueryItem *v, JsQueryItem *e);

void alignStringInfoInt(StringInfo buf);

/*
 * Parsing
 */

typedef struct JsQueryParseItem JsQueryParseItem;

struct JsQueryParseItem {
	JsQueryItemType	type;
	JsQueryParseItem	*next; /* next in path */

	union {
		struct {
			JsQueryParseItem	*left;
			JsQueryParseItem	*right;
		} args;

		JsQueryParseItem	*arg;

		Numeric		numeric;
		bool		boolean;
		struct {
			uint32      len;
			char        *val; /* could not be not null-terminated */
		} string;

		struct {
			int					nelems;
			JsQueryParseItem	**elems;
		} array;
	};
};

extern JsQueryParseItem* parsejsquery(const char *str, int len);

/* jsquery_extract.c */

typedef enum
{
	iAny 		= jqiAny,
	iAnyArray 	= jqiAnyArray,
	iKey 		= jqiKey,
	iAnyKey 	= jqiAnyKey
} PathItemType;

typedef struct PathItem PathItem;
struct PathItem
{
	PathItemType	type;
	int				len;
	char		   *s;
	PathItem	   *parent;
};

typedef enum
{
	eScalar = 1,
	eAnd	= jqiAnd,
	eOr		= jqiOr,
	eNot 	= jqiNot
} ExtractedNodeType;

typedef struct ExtractedNode ExtractedNode;
struct ExtractedNode
{
	ExtractedNodeType	type;
	PathItem		   *path;
	bool				indirect;
	union
	{
		struct
		{
			ExtractedNode **items;
			int				count;
		} args;
		struct
		{
			bool			inequality;
			bool			leftInclusive;
			bool			rightInclusive;
			JsQueryItem		*exact;
			JsQueryItem		*leftBound;
			JsQueryItem		*rightBound;
		} bounds;
		int	entryNum;
	};
};

typedef int (*MakeEntryHandler)(ExtractedNode *node, Pointer extra);

#define JsQueryMatchStrategyNumber			14

ExtractedNode *extractJsQuery(JsQuery *jq, MakeEntryHandler handler, Pointer extra);
bool queryNeedRecheck(ExtractedNode *node);
bool execRecursive(ExtractedNode *node, bool *check);
bool execRecursiveTristate(ExtractedNode *node, GinTernaryValue *check);

#endif
