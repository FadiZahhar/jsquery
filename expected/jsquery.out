CREATE EXTENSION jsquery;
set escape_string_warning=off;
CREATE TABLE test_jsquery (v jsonb);
\copy test_jsquery from 'data/test_jsquery.data'
select 'asd.zzz = 13'::jsquery;
     jsquery      
------------------
 "asd"."zzz" = 13
(1 row)

select 'asd.zzz < 13'::jsquery;
     jsquery      
------------------
 "asd"."zzz" < 13
(1 row)

select 'asd(zzz < 13)'::jsquery;
     jsquery      
------------------
 "asd"."zzz" < 13
(1 row)

select 'asd(zzz < 13 & x = true)'::jsquery;
            jsquery             
--------------------------------
 "asd"("zzz" < 13 & "x" = true)
(1 row)

select 'asd(zzz < 13 & x = "true")'::jsquery;
             jsquery              
----------------------------------
 "asd"("zzz" < 13 & "x" = "true")
(1 row)

select 'asd(zzz < 13 & x.zxc = "true")'::jsquery;
                jsquery                 
----------------------------------------
 "asd"("zzz" < 13 & "x"."zxc" = "true")
(1 row)

select 'asd(zzz < 13 | #.zxc = "true")'::jsquery;
               jsquery                
--------------------------------------
 "asd"("zzz" < 13 | #."zxc" = "true")
(1 row)

select 'asd(* < 13 & #.zxc = "true")'::jsquery;
             jsquery              
----------------------------------
 "asd"(* < 13 & #."zxc" = "true")
(1 row)

select '(* < 13 & #.zxc = "true")'::jsquery;
           jsquery           
-----------------------------
 (* < 13 & #."zxc" = "true")
(1 row)

select '* < 13 & #.zxc = "true"'::jsquery;
           jsquery           
-----------------------------
 (* < 13 & #."zxc" = "true")
(1 row)

select 'a < 1'::jsquery;
 jsquery 
---------
 "a" < 1
(1 row)

select 'a < -1'::jsquery;
 jsquery  
----------
 "a" < -1
(1 row)

select 'a < +1'::jsquery;
 jsquery 
---------
 "a" < 1
(1 row)

select 'a < .1'::jsquery;
  jsquery  
-----------
 "a" < 0.1
(1 row)

select 'a < -.1'::jsquery;
  jsquery   
------------
 "a" < -0.1
(1 row)

select 'a < +.1'::jsquery;
  jsquery  
-----------
 "a" < 0.1
(1 row)

select 'a < 0.1'::jsquery;
  jsquery  
-----------
 "a" < 0.1
(1 row)

select 'a < -0.1'::jsquery;
  jsquery   
------------
 "a" < -0.1
(1 row)

select 'a < +0.1'::jsquery;
  jsquery  
-----------
 "a" < 0.1
(1 row)

select 'a < 10.1'::jsquery;
  jsquery   
------------
 "a" < 10.1
(1 row)

select 'a < -10.1'::jsquery;
   jsquery   
-------------
 "a" < -10.1
(1 row)

select 'a < +10.1'::jsquery;
  jsquery   
------------
 "a" < 10.1
(1 row)

select 'a < 1e1'::jsquery;
 jsquery  
----------
 "a" < 10
(1 row)

select 'a < -1e1'::jsquery;
  jsquery  
-----------
 "a" < -10
(1 row)

select 'a < +1e1'::jsquery;
 jsquery  
----------
 "a" < 10
(1 row)

select 'a < .1e1'::jsquery;
 jsquery 
---------
 "a" < 1
(1 row)

select 'a < -.1e1'::jsquery;
 jsquery  
----------
 "a" < -1
(1 row)

select 'a < +.1e1'::jsquery;
 jsquery 
---------
 "a" < 1
(1 row)

select 'a < 0.1e1'::jsquery;
 jsquery 
---------
 "a" < 1
(1 row)

select 'a < -0.1e1'::jsquery;
 jsquery  
----------
 "a" < -1
(1 row)

select 'a < +0.1e1'::jsquery;
 jsquery 
---------
 "a" < 1
(1 row)

select 'a < 10.1e1'::jsquery;
  jsquery  
-----------
 "a" < 101
(1 row)

select 'a < -10.1e1'::jsquery;
  jsquery   
------------
 "a" < -101
(1 row)

select 'a < +10.1e1'::jsquery;
  jsquery  
-----------
 "a" < 101
(1 row)

select 'a < 1e-1'::jsquery;
  jsquery  
-----------
 "a" < 0.1
(1 row)

select 'a < -1e-1'::jsquery;
  jsquery   
------------
 "a" < -0.1
(1 row)

select 'a < +1e-1'::jsquery;
  jsquery  
-----------
 "a" < 0.1
(1 row)

select 'a < .1e-1'::jsquery;
  jsquery   
------------
 "a" < 0.01
(1 row)

select 'a < -.1e-1'::jsquery;
   jsquery   
-------------
 "a" < -0.01
(1 row)

select 'a < +.1e-1'::jsquery;
  jsquery   
------------
 "a" < 0.01
(1 row)

select 'a < 0.1e-1'::jsquery;
  jsquery   
------------
 "a" < 0.01
(1 row)

select 'a < -0.1e-1'::jsquery;
   jsquery   
-------------
 "a" < -0.01
(1 row)

select 'a < +0.1e-1'::jsquery;
  jsquery   
------------
 "a" < 0.01
(1 row)

select 'a < 10.1e-1'::jsquery;
  jsquery   
------------
 "a" < 1.01
(1 row)

select 'a < -10.1e-1'::jsquery;
   jsquery   
-------------
 "a" < -1.01
(1 row)

select 'a < +10.1e-1'::jsquery;
  jsquery   
------------
 "a" < 1.01
(1 row)

select 'a < 1e+1'::jsquery;
 jsquery  
----------
 "a" < 10
(1 row)

select 'a < -1e+1'::jsquery;
  jsquery  
-----------
 "a" < -10
(1 row)

select 'a < +1e+1'::jsquery;
 jsquery  
----------
 "a" < 10
(1 row)

select 'a < .1e+1'::jsquery;
 jsquery 
---------
 "a" < 1
(1 row)

select 'a < -.1e+1'::jsquery;
 jsquery  
----------
 "a" < -1
(1 row)

select 'a < +.1e+1'::jsquery;
 jsquery 
---------
 "a" < 1
(1 row)

select 'a < 0.1e+1'::jsquery;
 jsquery 
---------
 "a" < 1
(1 row)

select 'a < -0.1e+1'::jsquery;
 jsquery  
----------
 "a" < -1
(1 row)

select 'a < +0.1e+1'::jsquery;
 jsquery 
---------
 "a" < 1
(1 row)

select 'a < 10.1e+1'::jsquery;
  jsquery  
-----------
 "a" < 101
(1 row)

select 'a < -10.1e+1'::jsquery;
  jsquery   
------------
 "a" < -101
(1 row)

select 'a < +10.1e+1'::jsquery;
  jsquery  
-----------
 "a" < 101
(1 row)

select 'a in (0,1,2)'::jsquery;
     jsquery      
------------------
 "a" IN (0, 1, 2)
(1 row)

select 'a IN (0,null, "null", xxx, "zzz", 2)'::jsquery;
                  jsquery                  
-------------------------------------------
 "a" IN (0, null, "null", "xxx", "zzz", 2)
(1 row)

select '{"a": {"b": null}}'::jsonb @@ 'a.b = 1';
 ?column? 
----------
 f
(1 row)

select '{"a": {"b": null}}'::jsonb @@ 'a.b = null';
 ?column? 
----------
 t
(1 row)

select '{"a": {"b": null}}'::jsonb @@ 'a.b = false';
 ?column? 
----------
 f
(1 row)

select '{"a": {"b": false}}'::jsonb @@ 'a.b = false';
 ?column? 
----------
 t
(1 row)

select '{"a": {"b": false}}'::jsonb @@ 'a.b = true';
 ?column? 
----------
 f
(1 row)

select '{"a": {"b": true}}'::jsonb @@ 'a.b = true';
 ?column? 
----------
 t
(1 row)

select '{"a": {"b": 1}}'::jsonb @@ 'a.b = 1';
 ?column? 
----------
 t
(1 row)

select '{"a": {"b": 1}}'::jsonb @@ 'a.b < 1';
 ?column? 
----------
 f
(1 row)

select '{"a": {"b": 1}}'::jsonb @@ 'a.b <= 1';
 ?column? 
----------
 t
(1 row)

select '{"a": {"b": 1}}'::jsonb @@ 'a.b >= 1';
 ?column? 
----------
 t
(1 row)

select '{"a": {"b": 1}}'::jsonb @@ 'a.b > 1';
 ?column? 
----------
 f
(1 row)

select '{"a": {"b": 1}}'::jsonb @@ 'a.b = 2';
 ?column? 
----------
 f
(1 row)

select '{"a": {"b": 1}}'::jsonb @@ 'a.b < 2';
 ?column? 
----------
 t
(1 row)

select '{"a": {"b": 1}}'::jsonb @@ 'a.b <= 2';
 ?column? 
----------
 t
(1 row)

select '{"a": {"b": 1}}'::jsonb @@ 'a.b >= 2';
 ?column? 
----------
 f
(1 row)

select '{"a": {"b": 1}}'::jsonb @@ 'a.b > 2';
 ?column? 
----------
 f
(1 row)

select '{"a": {"b": 1}}'::jsonb @@ 'a.b = 0';
 ?column? 
----------
 f
(1 row)

select '{"a": {"b": 1}}'::jsonb @@ 'a.b < 0';
 ?column? 
----------
 f
(1 row)

select '{"a": {"b": 1}}'::jsonb @@ 'a.b <= 0';
 ?column? 
----------
 f
(1 row)

select '{"a": {"b": 1}}'::jsonb @@ 'a.b >= 0';
 ?column? 
----------
 t
(1 row)

select '{"a": {"b": 1}}'::jsonb @@ 'a.b > 0';
 ?column? 
----------
 t
(1 row)

select '{"a": {"b": 1}}'::jsonb @@ '*.b > 0';
 ?column? 
----------
 t
(1 row)

select '{"a": {"b": 1}}'::jsonb @@ '*.b > 0';
 ?column? 
----------
 t
(1 row)

select '{"a": {"b": 1}}'::jsonb @@ 'a.* > 0';
 ?column? 
----------
 t
(1 row)

select '{"a": {"b": 1}}'::jsonb @@ 'a.* > 0';
 ?column? 
----------
 t
(1 row)

select '{"a": {"b": [1,2,3]}}'::jsonb @@ '*.b && [ 1 ]';
 ?column? 
----------
 t
(1 row)

select '{"a": {"b": [1,2,3]}}'::jsonb @@ '*.b @> [ 1 ]';
 ?column? 
----------
 t
(1 row)

select '{"a": {"b": [1,2,3]}}'::jsonb @@ '*.b <@ [ 1 ]';
 ?column? 
----------
 f
(1 row)

select '{"a": {"b": [1,2,3]}}'::jsonb @@ '*.b @> [ 1,2,3,4 ]';
 ?column? 
----------
 f
(1 row)

select '{"a": {"b": [1,2,3]}}'::jsonb @@ '*.b <@ [ 1,2,3,4 ]';
 ?column? 
----------
 t
(1 row)

select '[{"a": 2}, {"a": 3}]'::jsonb @@ '*.a = 4';
 ?column? 
----------
 f
(1 row)

select '[{"a": 2}, {"a": 3}]'::jsonb @@ '*.a = 3';
 ?column? 
----------
 t
(1 row)

select '[{"a": 2}, {"a": 3}, {"a": {"a":4}}]'::jsonb @@ '#.a = 4';
 ?column? 
----------
 f
(1 row)

select '[{"a": 2}, {"a": 3}, {"a": {"a":4}}]'::jsonb @@ '*.a = 4';
 ?column? 
----------
 t
(1 row)

select '[{"a": 2}, {"a": 3}, {"a": {"a":4}}]'::jsonb @@ '#(a = 1 | a=3)';
 ?column? 
----------
 t
(1 row)

select '[{"a": 2}, {"a": 3}, {"a": {"a":4}}]'::jsonb @@ '#(a = 3 | a=1)';
 ?column? 
----------
 t
(1 row)

select '[{"a": 2}, {"a": 3}, {"a": {"a":4}}]'::jsonb @@ '#(a = 3 & a=1)';
 ?column? 
----------
 f
(1 row)

select '[{"a": 2}, {"a": 3}, {"a": {"a":4}}]'::jsonb @@ '#(a = 3 & a=2)' as "false";
 false 
-------
 f
(1 row)

select '[{"a": 2, "b":3}, {"a": 3, "b": 1}]'::jsonb @@ '#(b = 1 & a=3)';
 ?column? 
----------
 t
(1 row)

select '[{"a": 2}, {"a": 3}, {"a": {"a":4}}]'::jsonb @@ '#.a.a = 4';
 ?column? 
----------
 t
(1 row)

select '[{"a": 2}, {"a": 3}, {"a": {"a":4}}]'::jsonb @@ '*.a.a = 4';
 ?column? 
----------
 t
(1 row)

select '[{"a": 2}, {"a": 3}, {"a": {"a":4}}]'::jsonb @@ '*.#.a.a = 4';
 ?column? 
----------
 t
(1 row)

select '[{"a": 2}, {"a": 3}, {"a": {"a":4}}]'::jsonb @@ '#.*.a.a = 4';
 ?column? 
----------
 t
(1 row)

select '{"a": 1}'::jsonb @@ 'a in (0,1,2)';
 ?column? 
----------
 t
(1 row)

select '{"a": 1}'::jsonb @@ 'a in (0,2)';
 ?column? 
----------
 f
(1 row)

select '{"a": {"b": [1,2,3]}}'::jsonb @@ 'a.b.#=2';
 ?column? 
----------
 t
(1 row)

select '{"a": {"b": [1,2,3]}}'::jsonb @@ '*.b && [ 5 ]';
 ?column? 
----------
 f
(1 row)

select '{"a": {"b": [1,2,3]}}'::jsonb @@ 'a=*';
 ?column? 
----------
 t
(1 row)

select '{"a": {"b": [1,2,3]}}'::jsonb @@ 'a.b=*';
 ?column? 
----------
 t
(1 row)

select '{"a": {"b": [1,2,3]}}'::jsonb @@ 'a.c=*';
 ?column? 
----------
 f
(1 row)

select '{"a": {"b": [1,2,3]}}'::jsonb @@ 'a.b = [1,2,3]';
 ?column? 
----------
 t
(1 row)

select '{"a": {"b": [1,2,3]}}'::jsonb @@ 'a.b.# = [1,2,3]';
 ?column? 
----------
 f
(1 row)

select '{"a": {"b": [1,2,3]}}'::jsonb @@ 'a.b && [1,2,3]';
 ?column? 
----------
 t
(1 row)

select '{"a": {"b": [1,2,3]}}'::jsonb @@ 'a.b.# && [1,2,3]';
 ?column? 
----------
 f
(1 row)

select 'asd.# = 3'::jsquery & 'zzz = true' | 'xxx.# = zero';
                     ?column?                      
---------------------------------------------------
 (("asd".# = 3 & "zzz" = true) | "xxx".# = "zero")
(1 row)

select !'asd.# = 3'::jsquery & 'zzz = true' | !'xxx.# = zero';
                        ?column?                         
---------------------------------------------------------
 ((!("asd".# = 3) & "zzz" = true) | !("xxx".# = "zero"))
(1 row)

select '{"x":[0,1,1,2]}'::jsonb @@ 'x @> [1,0]'::jsquery;
 ?column? 
----------
 t
(1 row)

select '{"x":[0,1,1,2]}'::jsonb @@ 'x @> [1,0,1]'::jsquery;
 ?column? 
----------
 t
(1 row)

select '{"x":[0,1,1,2]}'::jsonb @@ 'x @> [1,0,3]'::jsquery;
 ?column? 
----------
 f
(1 row)

select '{"a": {"b": [1,2,3]}}'::jsonb @@ '*.b && [ 2 ]';
 ?column? 
----------
 t
(1 row)

select '{"a": {"b": [1,2,3]}}'::jsonb @@ '*.b($ && [ 2 ])';
 ?column? 
----------
 t
(1 row)

select '{"a": {"b": [1,2,3]}}'::jsonb @@ 'a.$.b && [ 2 ]';
 ?column? 
----------
 t
(1 row)

select '{"a": {"b": [1,2,3]}}'::jsonb @@ 'a.$.b ($ && [ 2 ])';
 ?column? 
----------
 t
(1 row)

select '[1,2,3]'::jsonb @@ '# && [2]';
 ?column? 
----------
 f
(1 row)

select '[1,2,3]'::jsonb @@ '#($ && [2])';
 ?column? 
----------
 f
(1 row)

select '[1,2,3]'::jsonb @@ '$ && [2]';
 ?column? 
----------
 t
(1 row)

select '[1,2,3]'::jsonb @@ '$ ($ && [2])';
 ?column? 
----------
 t
(1 row)

select '[1,2,3]'::jsonb @@ '$ = 2';
 ?column? 
----------
 f
(1 row)

select '[1,2,3]'::jsonb @@ '# = 2';
 ?column? 
----------
 t
(1 row)

select '[1,2,3]'::jsonb @@ '#.$ = 2';
 ?column? 
----------
 t
(1 row)

select '[1,2,3]'::jsonb @@ '#($ = 2)';
 ?column? 
----------
 t
(1 row)

select '{"a": {"b": 3, "c": "hey"}, "x": [5,6]}'::jsonb @@ '%.b=3';
 ?column? 
----------
 t
(1 row)

select '{"a": {"b": 3, "c": "hey"}, "x": [5,6]}'::jsonb @@ 'a.%=3';
 ?column? 
----------
 t
(1 row)

select '{"a": {"b": 3, "c": "hey"}, "x": [5,6]}'::jsonb @@ '%.%="hey"';
 ?column? 
----------
 t
(1 row)

select '{"a": {"b": 3, "c": "hey"}, "x": [5,6]}'::jsonb @@ '%="hey"';
 ?column? 
----------
 f
(1 row)

select '{"a": {"b": 3, "c": "hey"}, "x": [5,6]}'::jsonb @@ '%=[5,6]';
 ?column? 
----------
 t
(1 row)

select '"XXX"'::jsonb @@ '$="XXX"';
 ?column? 
----------
 t
(1 row)

select '"XXX"'::jsonb @@ '#.$="XXX"';
 ?column? 
----------
 f
(1 row)

---table and index
select count(*) from test_jsquery where (v->>'review_helpful_votes')::int4 > 0;
 count 
-------
   653
(1 row)

select count(*) from test_jsquery where (v->>'review_helpful_votes')::int4 > 19;
 count 
-------
    13
(1 row)

select count(*) from test_jsquery where (v->>'review_helpful_votes')::int4 < 19;
 count 
-------
   984
(1 row)

select count(*) from test_jsquery where (v->>'review_helpful_votes')::int4 >= 19;
 count 
-------
    16
(1 row)

select count(*) from test_jsquery where (v->>'review_helpful_votes')::int4 <= 19;
 count 
-------
   987
(1 row)

select count(*) from test_jsquery where (v->>'review_helpful_votes')::int4 = 19;
 count 
-------
     3
(1 row)

select count(*) from test_jsquery where (v->>'review_helpful_votes')::int4 > 16 AND
										(v->>'review_helpful_votes')::int4 < 20;
 count 
-------
     8
(1 row)

select count(*) from test_jsquery where v @@ 'review_helpful_votes > 0';
 count 
-------
   653
(1 row)

select count(*) from test_jsquery where v @@ 'review_helpful_votes > 19';
 count 
-------
    13
(1 row)

select count(*) from test_jsquery where v @@ 'review_helpful_votes < 19';
 count 
-------
   984
(1 row)

select count(*) from test_jsquery where v @@ 'review_helpful_votes >= 19';
 count 
-------
    16
(1 row)

select count(*) from test_jsquery where v @@ 'review_helpful_votes <= 19';
 count 
-------
   987
(1 row)

select count(*) from test_jsquery where v @@ 'review_helpful_votes = 19';
 count 
-------
     3
(1 row)

select count(*) from test_jsquery where v @@ 'review_helpful_votes > 16' AND
										v @@ 'review_helpful_votes < 20';
 count 
-------
     8
(1 row)

select count(*) from test_jsquery where v @@ 'review_helpful_votes > 16 & review_helpful_votes < 20';
 count 
-------
     8
(1 row)

select count(*) from test_jsquery where v @@ 'review_helpful_votes ($ > 16 & $ < 20)';
 count 
-------
     8
(1 row)

select count(*) from test_jsquery where v @@ 'similar_product_ids && ["0440180295"]';
 count 
-------
     7
(1 row)

select count(*) from test_jsquery where v @@ 'similar_product_ids(# = "0440180295") ';
 count 
-------
     7
(1 row)

select count(*) from test_jsquery where v @@ 'similar_product_ids.#($ = "0440180295") ';
 count 
-------
     7
(1 row)

select count(*) from test_jsquery where v @@ 'similar_product_ids && ["0440180295"] & product_sales_rank > 300000';
 count 
-------
     4
(1 row)

create index t_idx on test_jsquery using gin (v jsonb_bloom_value_ops);
set enable_seqscan = off;
explain (costs off) select count(*) from test_jsquery where v @@ 'review_helpful_votes > 0';
                               QUERY PLAN                               
------------------------------------------------------------------------
 Aggregate
   ->  Bitmap Heap Scan on test_jsquery
         Recheck Cond: (v @@ '"review_helpful_votes" > 0'::jsquery)
         ->  Bitmap Index Scan on t_idx
               Index Cond: (v @@ '"review_helpful_votes" > 0'::jsquery)
(5 rows)

select count(*) from test_jsquery where v @@ 'review_helpful_votes > 0';
 count 
-------
   653
(1 row)

select count(*) from test_jsquery where v @@ 'review_helpful_votes > 19';
 count 
-------
    13
(1 row)

select count(*) from test_jsquery where v @@ 'review_helpful_votes < 19';
 count 
-------
   984
(1 row)

select count(*) from test_jsquery where v @@ 'review_helpful_votes >= 19';
 count 
-------
    16
(1 row)

select count(*) from test_jsquery where v @@ 'review_helpful_votes <= 19';
 count 
-------
   987
(1 row)

select count(*) from test_jsquery where v @@ 'review_helpful_votes = 19';
 count 
-------
     3
(1 row)

select count(*) from test_jsquery where v @@ 'review_helpful_votes > 16' AND
										v @@ 'review_helpful_votes < 20';
 count 
-------
     8
(1 row)

select count(*) from test_jsquery where v @@ 'review_helpful_votes > 16 & review_helpful_votes < 20';
 count 
-------
     8
(1 row)

select count(*) from test_jsquery where v @@ 'review_helpful_votes ($ > 16 & $ < 20)';
 count 
-------
     8
(1 row)

select count(*) from test_jsquery where v @@ 'similar_product_ids && ["0440180295"]';
 count 
-------
     7
(1 row)

select count(*) from test_jsquery where v @@ 'similar_product_ids(# = "0440180295") ';
 count 
-------
     7
(1 row)

select count(*) from test_jsquery where v @@ 'similar_product_ids.#($ = "0440180295") ';
 count 
-------
     7
(1 row)

select count(*) from test_jsquery where v @@ 'similar_product_ids && ["0440180295"] & product_sales_rank > 300000';
 count 
-------
     4
(1 row)

drop index t_idx;
create index t_idx on test_jsquery using gin (v jsonb_hash_value_ops);
set enable_seqscan = off;
explain (costs off) select count(*) from test_jsquery where v @@ 'review_helpful_votes > 0';
                               QUERY PLAN                               
------------------------------------------------------------------------
 Aggregate
   ->  Bitmap Heap Scan on test_jsquery
         Recheck Cond: (v @@ '"review_helpful_votes" > 0'::jsquery)
         ->  Bitmap Index Scan on t_idx
               Index Cond: (v @@ '"review_helpful_votes" > 0'::jsquery)
(5 rows)

select count(*) from test_jsquery where v @@ 'review_helpful_votes > 0';
 count 
-------
   653
(1 row)

select count(*) from test_jsquery where v @@ 'review_helpful_votes > 19';
 count 
-------
    13
(1 row)

select count(*) from test_jsquery where v @@ 'review_helpful_votes < 19';
 count 
-------
   984
(1 row)

select count(*) from test_jsquery where v @@ 'review_helpful_votes >= 19';
 count 
-------
    16
(1 row)

select count(*) from test_jsquery where v @@ 'review_helpful_votes <= 19';
 count 
-------
   987
(1 row)

select count(*) from test_jsquery where v @@ 'review_helpful_votes = 19';
 count 
-------
     3
(1 row)

select count(*) from test_jsquery where v @@ 'review_helpful_votes > 16' AND
										v @@ 'review_helpful_votes < 20';
 count 
-------
     8
(1 row)

select count(*) from test_jsquery where v @@ 'review_helpful_votes > 16 & review_helpful_votes < 20';
 count 
-------
     8
(1 row)

select count(*) from test_jsquery where v @@ 'review_helpful_votes ($ > 16 & $ < 20)';
 count 
-------
     8
(1 row)

select count(*) from test_jsquery where v @@ 'similar_product_ids && ["0440180295"]';
 count 
-------
     7
(1 row)

select count(*) from test_jsquery where v @@ 'similar_product_ids(# = "0440180295") ';
 count 
-------
     7
(1 row)

select count(*) from test_jsquery where v @@ 'similar_product_ids.#($ = "0440180295") ';
 count 
-------
     7
(1 row)

select count(*) from test_jsquery where v @@ 'similar_product_ids && ["0440180295"] & product_sales_rank > 300000';
 count 
-------
     4
(1 row)

RESET enable_seqscan;
